src/XMakeBuildEngine/Definition/ToolsetReader.cs

```diff
@@ -205,42 +205,45 @@ ToolsetDefinitionLocations locations
                         // Other toolsets are installed in the xbuild directory
                         var xbuildToolsetsDir = Path.Combine(libraryPath, "xbuild");
                         var r = new Regex(xbuildToolsetsDir + Path.DirectorySeparatorChar + @"\d+\.\d+");
-                        foreach (var d in Directory.GetDirectories(xbuildToolsetsDir).Where(d => r.IsMatch(d)))
+                        if (Directory.Exists(xbuildToolsetsDir)) 
                         {
-                            var version = Path.GetFileName(d);
-                            var binPath = Path.Combine(d, "bin");
-                            if (version != null && !toolsets.ContainsKey(version))
+                            foreach (var d in Directory.GetDirectories(xbuildToolsetsDir).Where(d => r.IsMatch(d)))
                             {
-                                // Create standard properties. On Mono they are well known
-                                PropertyDictionary<ProjectPropertyInstance> buildProperties =
-                                    CreateStandardProperties(globalProperties, version, xbuildToolsetsDir, binPath);
+                                var version = Path.GetFileName(d);
+                                var binPath = Path.Combine(d, "bin");
+                                if (version != null && !toolsets.ContainsKey(version))
+                                {
+                                    // Create standard properties. On Mono they are well known
+                                    PropertyDictionary<ProjectPropertyInstance> buildProperties =
+                                        CreateStandardProperties(globalProperties, version, xbuildToolsetsDir, binPath);
+
+                                    toolsets.Add(
+                                        version,
+                                        new Toolset(
+                                            version,
+                                            binPath,
+                                            buildProperties,
+                                            environmentProperties,
+                                            globalProperties,
+                                            null,
+                                            currentDir,
+                                            string.Empty));
+                                }
+                            }
 
+                            if (!toolsets.ContainsKey(MSBuildConstants.CurrentToolsVersion))
```

Comment:
Does this `if` block need to be under the directory check? IIUC, this should be run irrespective of the earlier block.